chips/nrf52840/src/ieee802154_radio.rs

```diff
@@ -780,87 +803,109 @@ impl<'a> Radio<'a> {
                 self.registers.event_ready.write(Event::READY::CLEAR);
 
                 // Completed receiving a packet, now determine if we need to send ACK
                 if self.registers.event_end.is_set(Event::READY) {
                     self.registers.event_end.write(Event::READY::CLEAR);
-                    let result = self.crc_check();
+                    let crc = self.crc_check();
 
-                    // Unwrap fail = Radio RX Buffer is missing (may be due to receive client not replacing in receive(...) method,
-                    // or some instance in  driver taking buffer without properly replacing).
+                    // Unwrap fail = Radio RX Buffer is missing (may be due to
+                    // receive client not replacing in receive(...) method, or
+                    // some instance in driver taking buffer without properly
+                    // replacing).
                     let rbuf = self.rx_buf.take().unwrap();
 
-                    // data buffer format: | PSDU OFFSET | FRAME | LQI |
-                    // length of received data transferred to buffer (including PSDU)
-                    let data_len = rbuf[MIMIC_PSDU_OFFSET as usize] as usize;
-
-                    // Check if the received packet has a valid CRC and as a last resort
-                    // confirm that the received packet length field is in compliance
-                    // with the maximum packet length. If not, drop the packet.
-                    if !result.is_ok() || data_len >= MAX_FRAME_SIZE + PSDU_OFFSET {
-                        self.rx_buf.replace(rbuf);
-                        return
-                    }
-
-                    // lqi is found just after the data received
+                    // Data buffer format: | PREFIX | PHR | PSDU | LQI |
+                    //
+                    // Retrieve the length of the PSDU (actual frame). The frame
+                    // length is only 7 bits, but of course the field is a byte.
+                    // The nRF52840 product specification says this about the
+                    // PHR byte (Version 1.8, section 6.20.12.1):
+                    //
+                    // > The most significant bit is reserved and is set to zero
+                    // > for frames that are standard compliant. The radio
+                    // > module will report all eight bits and it can
+                    // > potentially be used to carry some information.
+                    //
+                    // We are not using that for any information so we just
+                    // force it to zero. This ensures that `data_len` will not
+                    // be longer than our buffer.
+                    let data_len = (rbuf[radio::PHR_OFFSET] & 0x7F) as usize;
+
+                    // LQI is found just after the data received.
                     let lqi = rbuf[data_len];
 
+                    // We drop the CRC bytes (the MFR) from our frame.
```

Comment:
Perhaps changing this to `buffer_len`? `buffer_len` too feels suboptimal.

This gets at the issue with needing this field to be passed up the stack to denote the active length since we are passing the entire unused buffer upwards (so that it can later be reset by framer). These issues all seem like they would be solved by using a `subslice`. Is there a reason that we did not originally do this? In my opinion it seems the radio driver itself should be the only entity setting/resetting the receive buffer it interacts with (not the framer or some other "higher" element).

Reply:
> This gets at the issue with needing this field to be passed up the stack to denote the active length since we are passing the entire unused buffer upwards (so that it can later be reset by framer). These issues all seem like they would be solved by using a `subslice`. Is there a reason that we did not originally do this? In my opinion it seems the radio driver itself should be the only entity setting/resetting the receive buffer it interacts with (not the framer or some other "higher" element).

The first 15.4 code was probably written before subslice/leasablebuffer existed.

I agree this should use subslice. But, the stack needs to edit both the start and end of the buffer, and that is more complex than many use cases. It sounded like the networking WG was thinking about that.

I would like to see the stack be clarified, and then updated to subslice once we have a good story for reserving space at the start of the buffer.